### PR TITLE
Extensibility tests: Lifetime - JWT, SAML and SAML2

### PR DESCRIPTION
**Pull request by iNinja**
# Extensibility tests: Lifetime - JWT, SAML and SAML2
- Handle potential exception thrown by the lifetime validation delegate in JWT, SAML, and SAML2
- Added custom lifetime validation delegates for testing
- Added lifetime extensibility tests for JWT, SAML, and SAML2

Part of #2711

### CODE REVIEW

**Review by pmaytak — APPROVED** [review]

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs [hunk 1]
```diff
@@ -255,13 +255,28 @@ private async ValueTask<ValidationResult<ValidatedToken>> ValidateJWSAsync(
             DateTime? expires = jsonWebToken.HasPayloadClaim(JwtRegisteredClaimNames.Exp) ? jsonWebToken.ValidTo : null;
             DateTime? notBefore = jsonWebToken.HasPayloadClaim(JwtRegisteredClaimNames.Nbf) ? jsonWebToken.ValidFrom : null;
 
-            ValidationResult<ValidatedLifetime> lifetimeValidationResult = validationParameters.LifetimeValidator(
-                notBefore, expires, jsonWebToken, validationParameters, callContext);
+            ValidationResult<ValidatedLifetime> lifetimeValidationResult;
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:258]: Since this entire block has grown, can you move it to a separate method to clean this area up?
```cs
ValidationResult<ValidatedLifetime> lifetimeValidationResult = ValidateLifetime(...);

// Then ValidateLifetime has the try-catch and handles the actual validation
```

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateToken.Internal.cs [hunk 2]
```diff
@@ -162,17 +162,32 @@ internal virtual ValidationResult<ValidatedConditions> ValidateConditions(
                     StackFrames.AssertionConditionsNull);
             }
 
-            var lifetimeValidationResult = validationParameters.LifetimeValidator(
-                samlToken.Assertion.Conditions.NotBefore,
-                samlToken.Assertion.Conditions.NotOnOrAfter,
-                samlToken,
-                validationParameters,
-                callContext);
+            ValidationResult<ValidatedLifetime> lifetimeValidationResult;
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateToken.Internal.cs:195]: Same thing here with moving this to a separate method.

File: src/Microsoft.IdentityModel.Tokens/InternalAPI.Unshipped.txt [hunk 3]
```diff
@@ -1,19 +1,23 @@
 const Microsoft.IdentityModel.Tokens.LogMessages.IDX10002 = "IDX10002: Unknown exception type returned. Type: '{0}'. Message: '{1}'." -> string
 const Microsoft.IdentityModel.Tokens.LogMessages.IDX10268 = "IDX10268: Unable to validate audience, validationParameters.ValidAudiences.Count == 0." -> string
 const Microsoft.IdentityModel.Tokens.LogMessages.IDX10269 = "IDX10269: IssuerValidationDelegate threw an exception, see inner exception." -> string
+const Microsoft.IdentityModel.Tokens.LogMessages.IDX10271 = "IDX10271: LifetimeValidationDelegate threw an exception, see inner exception." -> string
 const Microsoft.IdentityModel.Tokens.LogMessages.IDX10272 = "IDX10272: SignatureValidationDelegate threw an exception, see inner exception." -> string
 const Microsoft.IdentityModel.Tokens.LogMessages.IDX10273 = "IDX10273: AlgorithmValidationDelegate threw an exception, see inner exception." -> string
-const Microsoft.IdentityModel.Tokens.LogMessages.IDX10274 = "IDX10274: IssuerSigningKeyValidationDelegate threw an exception, see inner exception." -> string
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens/InternalAPI.Unshipped.txt:7]: What are all these seemingly unrelated changes?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+            try
+            {
+                ValidationResult<ValidatedToken> validationResult = await theoryData.JsonWebTokenHandler.ValidateTokenAsync(
+                    theoryData.JsonWebToken!,
+                    theoryData.ValidationParameters!,
+                    theoryData.CallContext,
+                    CancellationToken.None);
+
+                if (validationResult.IsValid)
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: Is this ever `true`? It seems all the test cases are negative tests. If so, can you add a positive test case?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+                {
+                    ValidatedToken validatedToken = validationResult.UnwrapResult();
+                    if (validatedToken.ValidatedLifetime is not null)
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: If the result is successful, then I think this should always be non-`null` for this test case. Therefore, this should be an assertion.
```cs
Assert.NotNull(validatedToken.ValidatedLifetime);
```
or
```cs
Assert.True(validatedToken.ValidatedLifetime.HasValue);
```

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+                        IdentityComparer.AreValidatedLifetimesEqual((ValidatedLifetime)validatedToken.ValidatedLifetime, theoryData.ValidatedLifetime, context);
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]:
```suggestion
                        IdentityComparer.AreValidatedLifetimesEqual(validatedToken.ValidatedLifetime.Value, theoryData.ValidatedLifetime, context);
```

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs [hunk 7]
```diff
@@ -255,13 +255,28 @@ private async ValueTask<ValidationResult<ValidatedToken>> ValidateJWSAsync(
...
 
-            if (!lifetimeValidationResult.IsValid)
+            try
+            {
+                lifetimeValidationResult = validationParameters.LifetimeValidator(
+                    notBefore, expires, jsonWebToken, validationParameters, callContext);
+
+                if (!lifetimeValidationResult.IsValid)
+                    return lifetimeValidationResult.UnwrapError().AddCurrentStackFrame();
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:266]: This is confusing. Can you add a method `ToResult()` and explicitly call it instead of using an `implicit` conversion?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs (hunk 4 above)
Comment (msbw2) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: Confirmed that this code path is never exercised.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+                }
+                else
+                {
+                    ValidationError validationError = validationResult.UnwrapError();
+                    IdentityComparer.AreValidationErrorsEqual(validationError, theoryData.LifetimeValidationError, context);
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: Can inlining cause any issues when stack frames/traces are compared in this assertion?

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandler.Extensibility.Lifetime.cs [hunk 9]
```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandler.Extensibility.Lifetime.cs:1]: Same comments in other test class apply here.

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/SamlSecurityTokenHandler.Extensibility.Lifetime.cs [hunk 10]
```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.Tokens.Saml.Tests/SamlSecurityTokenHandler.Extensibility.Lifetime.cs:1]: Same comments in other test class apply here.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs [hunk 11]
```diff
@@ -168,17 +168,32 @@ internal virtual ValidationResult<ValidatedConditions> ValidateConditions(
                     StackFrames.AssertionConditionsNull);
             }
 
-            var lifetimeValidationResult = validationParameters.LifetimeValidator(
-                samlToken.Assertion.Conditions.NotBefore,
-                samlToken.Assertion.Conditions.NotOnOrAfter,
-                samlToken,
-                validationParameters,
-                callContext);
+            ValidationResult<ValidatedLifetime> lifetimeValidationResult;
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs:198]: Same comments apply here.

File: src/Microsoft.IdentityModel.Tokens/InternalAPI.Unshipped.txt (hunk 3 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/InternalAPI.Unshipped.txt:7]: I think this was caused by the constant branch switch between all the extensibility test branches and Visual Studio's autogeneration of this file.
I believe when I was using certain branches as base, a line in the Unshipped.txt file remained when it was not in use, and now it is removing it.

Once all the extensibility test branches are merged, all of these will be present.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs (hunk 4 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: I have started to change this to a failure if it enters, since all our extensibility tests are around the errors returned from the custom delegates. I'll update this one as well.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs (hunk 7 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:266]: The idea behind the result type is that is a **one of**, in this case a **ValidatedToken** or a **ValidationError**.
We make use of the implicit conversion in order to allow the user to precisely return either of those.

I believe that doing `return new (lifetimeValidationResult.UnwrapError().AddCurrentStackFrame());` defeats that purpose to some extent.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs (hunk 1 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ValidateToken.Internal.cs:258]: I'll make a note to review this once all the validations are merged.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateToken.Internal.cs (hunk 2 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateToken.Internal.cs:195]: ack

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs (hunk 5 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.Lifetime.cs]: Changed as part of a separate feedback.

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandler.Extensibility.Lifetime.cs (hunk 9 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandler.Extensibility.Lifetime.cs:1]: ack

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/SamlSecurityTokenHandler.Extensibility.Lifetime.cs (hunk 10 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Saml.Tests/SamlSecurityTokenHandler.Extensibility.Lifetime.cs:1]: ack

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs (hunk 11 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs:198]: ack

**Review by westin-m — APPROVED** [review]

**Review by kllysng — APPROVED** [review]
